### ISSUE TITLE
[feature] Improve caching when tasks do not want to consider outputs

### ISSUE THREAD

**Issue by aidanhs** [moonrepo/moon#638]
**Is your feature request related to a problem? Please describe.**

Given the following `abc/moon.yml`:
```
tasks:
  test1:
    #command: "bash -c 'mkdir -p out && echo hi > out/a'"
    #command: "bash -c 'mkdir -p out && echo hii > out/a'"
    #command: "bash -c 'mkdir -p out && echo hi > out/a'"
    platform: 'system'
    outputs: [out]
```

If I run this task three times *uncommenting each `command` in turn) I get no cache hits.

This is expected behaviour - as documented on https://moonrepo.dev/docs/concepts/cache, apparently `output` targets are hashed before considering caches. But for build processes that always overwrite output/replace the output directory, this is overly pessimistic and means I will almost never get cache hits while iterating on some code or switching back and forward between different versions of something.

**Describe the solution you'd like**

I'd like a way to indicate to moon that the outputs should not be considered for caching purposes - perhaps this means moon considers them empty (and then clears them before running the task), or something else.

**Describe alternatives you've considered**

**Additional context**

**Comment by milesj** [comment]
@aidanhs Ok so I was able to reproduce this, but after looking at the hash contents, I know why this is happening, for 2 reasons:

- The first is that the "out" folder on subsequent runs is included as an input, since inputs default to `**/*`. I suppose this is a bug? I'm not entirely sure what the default should be here.
- The second is that the hash of `moon.yml` changes if you are simply removing the `#` from each command, but keeping all 3 commands. This is causing a different content hash each time.

<img width="1281" alt="Screenshot 2023-02-21 at 1 20 28 PM" src="https://user-images.githubusercontent.com/143744/220461315-afefdd53-b858-40bb-9cf2-42c734dd3ba5.png">

**Comment by aidanhs** [comment]
Oh that's very interesting! And the only reason moon.yml is being considered is because of that default inputs of `**/*` right? I can see that if I add a single explicit input then the third command does indeed get cached.

Which means I can now reproduce the reason I was paranoid about the caching in the first place! See #640 

I think that default for inputs is probably sane...trying to exclude outputs directories sounds interesting, but could get tricky in the presence of symlinks/hardlinks. I reckon you could probably close this out by adding a warning on the cache concepts page (https://moonrepo.dev/docs/concepts/cache) that inputs is important because of the default of `**/*` and also disable caching if inputs is left as the default (could put a debug log in case people want to dig into why caching doesn't happen).

**Comment by milesj** [comment]
Took a stab at this here: https://github.com/moonrepo/moon/pull/641 It's tricky...

Adding a note to the docs is a good idea though, will do.

**Comment by milesj** [comment]
Thanks for the issue. This landed in the 0.25 release. Outputs are now ignored as input sources, and I've updated the cache docs with some info on this.
